core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java

```diff
@@ -120,12 +120,7 @@ public void write(T row) throws IOException {
       // Create a copied key from this row.
       StructLike copiedKey = StructCopy.copy(structProjection.wrap(asStructLike(row)));
 
-      // Adding a pos-delete to replace the old path-offset.
-      PathOffset previous = insertedRowMap.put(copiedKey, pathOffset);
-      if (previous != null) {
-        // TODO attach the previous row if has a positional-delete row schema in appender factory.
-        posDeleteWriter.delete(previous.path, previous.rowOffset, null);
-      }
+      insertedRowMap.put(copiedKey, pathOffset);
```

Comment:
Why is this changing as well? I think that the logic here was intended to implement upsert just in case there are duplicate inserts without a delete. @openinx, can you take a look at this?

Reply:
@hameizi can you reply to this comment?

Reply:
@rdblue yes,as you said that the logic here was intended to implement upsert just in case there are duplicate inserts without a delete. This logic can be retain, just i think duplicate inserts should be avoid by upstream semantic but not writer. If need i will rollback this change.

Reply:
@rdblue After test i think this is necessary, and i think old logic is error. Because old logic only write pos-delete in `write` function what should happen in `delete` function. And below test case is also puzzle:
(https://github.com/apache/iceberg/blob/9b6b5e0d2e760694e2abef73fa9036d1d8bbd014/data/src/test/java/org/apache/iceberg/io/TestTaskEqualityDeltaWriter.java#L324), this test write duplicate key but return just one data but user should avoid write duplicate key depend on config `write.upsert.enable` https://github.com/apache/iceberg/pull/2863. So this PR move pos-delete logic in write function to delete function.

Reply:
Let's remove this change.

The logic here is to opportunistically catch duplicates when there are only inserts. This is not intended to replace the real upsert logic, which requires calling delete as you noted. Instead, it is here because we're updating the `insertedRowMap` and may get a previous insert location. When that happens, the right thing to do is to delete the duplicate row.

I also just realized that the changes below are incorrect. Instead of calling `internalPosDelete(key)`, this checks the `insertedRowMap` itself using `get`. The logic in internalPosDelete` used `remove` so that the entry was removed _before_ the insert occurred and we have a second opportunistic check.

To fix this, you should instead update `internalPosDelete` to return `true` if a row was deleted and `false` otherwise. Then you can update your `previous != null` check like this:

```java
    public void delete(T row) throws IOException {
      if (!internalPosDelete(structProjection.wrap(asStructLike(row)))) {
        eqDeleteWriter.write(row);
      }
    }
```